### PR TITLE
feat: Typescript 2/4

### PR DESCRIPTION
**Pull request by jimmycallin**
Partially resolves FT-0403674e-a5a4-11ed-96fe-bef30e7c0f4c

- [x] I have added automatic tests where applicable
- [x] The PR title is suitable as a release note
- [x] The PR contains a description of what has been changed
- [x] The description contains manual test instructions

## Changes

Step 2 of 4 in Typescript migration. Builds upon 1. Migrates minor files to ts.

## Test

Test against another codebase (preferably with typescript support) and make sure it builds and runs correctly.

### CODE REVIEW

File: source/util/encode_uri_parameters.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // :copyright: Copyright (c) 2018 ftrack
 
-function encodeUriParameters(data) {
```
Comment (WaylonFtrack) [inline source/util/encode_uri_parameters.ts:3]: I think remove it and to use `URLSearchParams`

File: source/event.ts [hunk 2]
```diff
@@ -5,14 +5,24 @@ import { v4 as uuidV4 } from "uuid";
...
+
   /**
    * Construct Event instance with *topic*, *data* and additional *options*.
    *
    * *topic* should be a string representing the event.
    *
    * *data* should be an object with the event payload.
    */
-  constructor(topic, data, options = {}) {
+  constructor(topic: string, data: any, options: { [key: string]: any } = {}) {
```
Comment (WaylonFtrack) [inline source/event.ts:25]: some of `any` on input type can be replaced by `object`, which is slightly more accurate.
like the `data` and `options` in here.

**Review by WaylonFtrack — CHANGES_REQUESTED** [review]

File: source/util/encode_uri_parameters.ts (hunk 1 above)
Comment (jimmycallin) [reply source/util/encode_uri_parameters.ts]: fixed!

File: source/event.ts (hunk 2 above)
Comment (jimmycallin) [reply source/event.ts]: fixed!

File: source/operation.ts [hunk 3]
```diff
@@ -4,6 +4,8 @@
  * @namespace operation
  */
 
+export interface Operation {}
```
Comment (gismya) [inline source/operation.ts:7]: Expand this one to add `action` and add separate interfaces for the different operation types

File: source/project_schema.ts [hunk 4]
```diff
@@ -78,32 +79,33 @@ export function getStatuses(
...
+      const schema = session.getSchema(entityType) as any;
 
       if (schema && schema.alias_for && schema.alias_for.id === "Task") {
         const objectTypeId = schema.alias_for.classifiers.object_typeid;
 
         for (const index in data._schemas) {
           if (data._schemas[index].type_id === objectTypeId) {
             statuses = data._schemas[index].statuses.map(
-              (status) => status.task_status
+              (status: any) => status.task_status
```
Comment (gismya) [inline source/project_schema.ts:108]: Remove this any typing in PR 4

File: source/operation.ts [hunk 5]
```diff
@@ -4,6 +4,46 @@
  * @namespace operation
  */
 
+export interface CreateOperation {
```
Comment (gismya) [inline source/operation.ts:7]: Add them as return types

File: source/operation.ts (hunk 3 above)
Comment (jimmycallin) [reply source/operation.ts]: fixed!

File: source/operation.ts (hunk 5 above)
Comment (jimmycallin) [reply source/operation.ts:7]: fixed!

File: source/project_schema.ts (hunk 4 above)
Comment (jimmycallin) [reply source/project_schema.ts:108]: ok!

**Review by gismya — APPROVED** [review]
Looking good!

**Review by CzBiX — APPROVED** [review]

**Review by WaylonFtrack — APPROVED** [review]

**Pull request merged**